streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala

```diff
@@ -40,6 +43,8 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
   private var totalProcessedRecords = 0L
   private val receiverInfos = new HashMap[Int, ReceiverInfo]
 
+  private val batchTimeToBatchUIData = new HashMap[Time, BatchUIData]
```

Comment:
> So we should convert all (waiting/running/completed)BatchInfos to ***BatchUIData. Then we dont need to add this extra batchTimeToBatchUIData any more.

It's a bit complex here. Because `onJobStart` and `onBatchXXX` messages are processed in different threads, we may not be able to get the corresponding BatchInfo in `onJobStart`. In such case, if we convert all `(waiting/running/completed)BatchInfos` to `***BatchUIData`, we still need to have an extra data structure to store `BatchUIData`s that does not have a corresponding `BatchInfo`.